drivers: uart_nrfx_uart: Replace k_delayed_work with k_timer

anangl opened the pull request:
There is no need to involve a work queue to control timeouts in
the uart_nrfx_uart driver. Kernel timers are sufficient for those
tasks, and the uart_nrfx_uarte driver may serve as a proof of this.
Replace then those uses of work queue with timers, for consistency
and simplicity, also to avoid potential issues with cancellation of
the delayed work queue items.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: uart_nrfx_uart: Replace k_delayed_work with k_timer

There is no need to involve a work queue to control timeouts in
the uart_nrfx_uart driver. Kernel timers are sufficient for those
tasks, and the uart_nrfx_uarte driver may serve as a proof of this.
Replace then those uses of work queue with timers, for consistency
and simplicity, also to avoid potential issues with cancellation of
the delayed work queue items.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

pabigot commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -597,14 +597,15 @@ static void rx_isr(const struct device *dev)
-					      K_MSEC(uart0_cb.rx_timeout));
+			k_timer_start(&uart0_cb.rx_timeout_timer,
+				      K_MSEC(uart0_cb.rx_timeout),
+				      K_MSEC(uart0_cb.rx_timeout));
```
Why  are these repeating rather than one-shot (third argument `K_NO_WAIT`)?  Doesn't this cause the handler to be invoked repeatedly even when there's no data?

anangl replied:
My bad. I blindly copied that from `uart_nrfx_uarte.c` where this timer is used in a slightly different way (there it is supposed to be invoked repeatedly). Thanks for catching this!

anangl replied:
Fixed. I extended also a bit the `test_single_read` test case so that mistakes like this could be caught there.

pabigot reviewed (changes requested):
The use of automatically-repeating timers seems wrong, certainly not what happens with the delayed-work solution.

anangl committed:
tests: uart_async_api: Catch extra RX_RDY events in test_single_read

Extend the test_single_read test case to catch RX_RDY events that are
not expected to appear.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

pabigot reviewed (approved):
Haven't tested but by inspection this looks good.

nordic-krch commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -640,8 +641,9 @@ static void tx_isr(const struct device *dev)
-					      K_MSEC(uart0_cb.tx_timeout));
+			k_timer_start(&uart0_cb.tx_timeout_timer,
+				      K_MSEC(uart0_cb.tx_timeout),
+				      K_MSEC(uart0_cb.tx_timeout));
```
shouldn't here also be `K_NO_WAIT`?

nordic-krch replied:
also this timeout i handled differently than the one in uarte. Here timeout is reset on each byte received and in uarte it is set once before starting.

anangl replied:
> shouldn't here also be `K_NO_WAIT`?

Of course it should. And it is so in `uart_nrfx_uarte.c`. How did I manage to copy this code incorrectly? Shame on me.

anangl replied:
Fixed.

anangl replied:
> also this timeout i handled differently than the one in uarte. Here timeout is reset on each byte received and in uarte it is set once before starting.

Yes. The timer should be rather started in `uart_nrfx_tx()`, not in `tx_isr()` before sending consecutive bytes. The description of API function `uart_tx()` does not state it explicitly, but it seems reasonable to use the timeout value for the whole transaction, not between individual bytes. And [this](https://github.com/zephyrproject-rtos/zephyr/blob/e7df25c848c6ab54bc81430605454f91ca204623/include/drivers/uart.h#L92-L93) note seems to be confirming that.
But I'm not sure if I should try to correct it in this PR.

nordic-krch replied:
imo, should be fixed separately as this PR focuses on replacing k_delayed_work
